### PR TITLE
Manylinux2010

### PR DESCRIPTION
**Pull request by dolang**
Now that there's some activity again here, I figure it's easier to expose the changes I made as a pull request so you can have a look at the diffs.

This is from my `manylinux2010` branch which is based on `markrwilliams:manylinx2` merged into `pypa:master`.

I tried to keep this branch close to upstream and just merged the most recent commit to master (4a20e4b) with minor adjustments.

I don't expect this to be merged by itself, but hope it saves you guys some time.  Notable:
- Removed `gpg` from `PYTHON_COMPILE_DEPS` because it's already present on CentOS 6 and its erasure later in the script breaks the build.
- Removed `ncurses-devel` as required by the PEP.
- Added `perl-devel`, seems to be required for the build now.
- Added a workaround for `libcurl.so` incompatibility (see commit notes & comments).
- Removed the `get-pip.py` workaround. That has been fixed (see [here](https://github.com/pypa/get-pip/issues/19)) and my build script works again.
- Moved `libffi-devel` from `PYTHON_COMPILE_DEPS` to the *"Development tools and libraries"* section.
- Changed the `x86_64` Docker tag from `:6.9` to `:6`. This mirrors a `manylinux1` change intended to reduce the overall image size (see pypa#206 and also [:6 Docker branch](https://github.com/CentOS/sig-cloud-instance-images/tree/CentOS-6/docker) / [:6.9 Docker branch](https://github.com/CentOS/sig-cloud-instance-images/tree/CentOS-6.9/docker)).
- Moved from CentOS 6.9 to 6.10
- **Careful:** I also changed occurrences of `manylinux1` to either `manylinux` or `manylinux2010`. This might be undesired.

~Side note: not a fan of the `/dev/null` out/error stream redirections in the scripts. They just make it harder to debug when something fails.  And there definitely are moving parts, as the `get-pip.py` problem has shown.~

### CODE REVIEW

**Comment by njsmith** [comment]
Unfortunately, some amount of `> /dev/null` redirections are mandatory because Travis unconditionally kills any job that prints too much output.

**Comment by dolang** [comment]
Good thing you mentioned that (I didn't actually know).  While I've removed those from my master, I actually tried to not mess with this `manylinx2010` branch. 

I've just noticed that I've removed two of those in this branch and you're making a very good point for keeping them in.  So I'll add them back.

An alternative would be to redirect to e.g. `>> /var/log/build.log` and remove that file again during the cleanup step.  In case the build breaks (locally), one could at least `docker commit` to inspect the log instead of rerunning it manually.  But it's a minor issue anyway.

Just noticed as well that the diffs are not that great because I merged the latest commits into this preexisting branch. You'd have to rebase/merge locally to get a better result, I guess. Sorry about that.

**Comment by dolang** [comment]
Note: The [travis CI build](https://github.com/pypa/manylinux/pull/182) does actually run now for this pull request, but despite what it says, it didn't complete successfully.

The newly added [build.sh](https://github.com/pypa/manylinux/blob/master/build.sh) needs a `set -e -x` or something like that.

I run a local build before each merge commit, however. So it should be fine.

**Comment by trishankatdatadog** [comment]
Sorry about the new build.sh, @dolang, that was me. Want me to add those shell options?

**Comment by dolang** [comment]
Myself, I don't need it because I don't (yet) use travis for anything. I have a small build script in my master branch. But I guess it would be in your interest to add those options (if travis allows them)? In any case, I can't verify that the code for travis is correct.

I'm not really in the loop here, I'm just keeping this pull request up-to-date as a reference/help for you guys once it's time to make the switch to manylinux2010.

**Comment by trishankatdatadog** [comment]
I think I see the problem now. Hopefully fixed in #190...

**Comment by dolang** [comment]
`x86_64` still fails, but that's not a surprise because it depends on that third Docker image which is not required for `manylinux1`.  I might try to fix that later.

However, `i686` seems to have completed a full build, so #190 definitely helped.

**Comment by trishankatdatadog** [comment]
Glad to hear, let me know if I can help with anything else!

**Comment by dolang** [comment]
Took a stab at it, but I'm not sure if the `no-vsyscall` Docker image should eventually get the `$TRAVIS_COMMIT` tag, too. So I left it out because that's easier.

**Comment by dolang** [comment]
I had to move `libffi-devel` from the `PYTHON_COMPILE_DEPS` because it can't be removed after the build, breaking the script:
```
install-info: No such file or directory for /usr/share/info/libffi.info.gz
error: %preun(libffi-devel-3.0.5-3.2.el6.x86_64) scriptlet failed, exit status 1
Error in PREUN scriptlet in rpm package libffi-devel
libffi-devel-3.0.5-3.2.el6.x86_64 was supposed to be removed but is not!
  Verifying  : libffi-devel-3.0.5-3.2.el6.x86_64

Failed:
  libffi-devel.x86_64 0:3.0.5-3.2.el6
```

It's only 21 Kb in size, so keeping it in the *"Development tools and libraries"* section is not a big problem anyway.

**Comment by dolang** [comment]
Well, that's odd. It did build on my machine before committing, though. From the Travis log:
```
$ PLATFORM=$PLATFORM TRAVIS_COMMIT=$TRAVIS_COMMIT travis_wait 60 ./build.sh


Still running (49 of 60): ./build.sh

The job exceeded the maximum time limit for jobs, and has been terminated.
```
` Ran for 49 min 29 sec`

**Comment by dolang** [comment]
On another note: Looks like [CentOS 6.10 is available now](http://vault.centos.org/), and I should look through the code again.

I'll have a look at that later.

**Comment by dolang** [comment]
Somehow the Travis build stops at 50 minutes although `travis_wait` is set to 60. Locally, the images build fine.

I've had a look around, maybe I should try the workaround linked [here](https://github.com/travis-ci/travis-ci/issues/8575#ref-commit-0dbb14c) from [this commit](https://github.com/cyberFund/cybernode-archive/commit/0dbb14c5169144b7535cbf3dc474916b93a64a5e)?

**Comment by dolang** [comment]
I've been able to move to CentOS 6.10 now, previously some repository information was missing (namely the `repodata` subfolder in http://vault.centos.org/6.10/os/Source/).

There's a new build of `glibc` in there, so I had to update the version from 1.209 to 1.212. Now here's a question regarding the custom `glibc` version:

I think I've run into a problem at some point because the new version number that is assigned in the `no-vsyscall` simply appends a `.1` (dot one),  i.e. now it's `1.212.1`, but I don't quite remember because it's already been a few months. To someone more familiar with RPMs: Wouldn't it be better to append a `-1` (dash one) denoting a new build instead of a new version?

**Comment by njsmith** [comment]
I can explain the timeout stuff anyway... Travis puts a hard limit of 50 minutes on all (open source) build jobs: https://docs.travis-ci.com/user/customizing-the-build/#build-timeouts

Travis also kills jobs that spend 10 minutes without producing output; the `travis_wait` script is a hack to avoid hitting this 10 minute timeout, by producing a continuous stream of meaningless output. But it doesn't help with the 50 minute limit.

**Comment by jcfr** [comment]
@njsmith  Would you consider switching to CircleCI ?  I would be happy to help.

**Comment by njsmith** [comment]
@jcfr The time I have to spend on manylinux-related things has gone down a bunch in the last few months, and I was kind of the only one doing things. So, honestly as far as I'm concerned, anyone who wants to step up and do the work is pretty much welcome to do whatever works for them.

**Comment by dralley** [comment]
Bump: Is this work going to proceed any time soon?

**Comment by mayeut** [comment]
If I'm not mistaken, auditwheel still needs to get updated. c.f. #179

**Comment by trishankatdatadog** [comment]
@mayeut Does that require another PR here?

**Comment by mayeut** [comment]
Once audit wheel gets released - I have no idea when, I commented on a PR there some time ago -, the pyup-bot will open a PR that needs to get merged in master (if CI is ok of course). Then I guess @dolang will have to merge upstream changes once more in this PR.
After that, I think it should be ok to merge if reviewers have nothing more to say. I guess manylinux1 (current master) will still have to be maintained for some time in a maintenance branch.

**Comment by trishankatdatadog** [comment]
Ok, sounds good. I'll try to read up on the PEP and this PR in the meantime.

File: docker/build_scripts/build.sh [hunk 1]
```diff
@@ -43,16 +33,17 @@ yum -y update
 
 # EPEL support
 yum -y install wget
-# https://dl.fedoraproject.org/pub/epel/5/x86_64/epel-release-5-4.noarch.rpm
-cp $MY_DIR/epel-release-5-4.noarch.rpm .
-check_sha256sum epel-release-5-4.noarch.rpm $EPEL_RPM_HASH
+# https://dl.fedoraproject.org/pub/epel/6/x86_64/epel-release-6-8.noarch.rpm
+cp $MY_DIR/epel-release-6-8.noarch.rpm .
```
Comment (galv) [inline docker/build_scripts/build.sh]: I don't think we should be concerned about this, but note that this doesnot correspond to centos 6.10. I'm not sure what the canonical location for the extended package for enterpriselinux (EPEL) rpm is, myself.

File: docker/glibc/README.rst [hunk 2]
```diff
@@ -0,0 +1,79 @@
+centos-6-no-vsyscall
+====================
+
+*Summary*: Because of
+https://mail.python.org/pipermail/wheel-builders/2016-December/000239.html,
+this a CentOS 6.10 Docker image that rebuilds ``glibc`` without
+*vsyscall* is necessary to reliably run ``manylinux2`` on 64-bit
```
Comment (galv) [inline docker/glibc/README.rst]: manylinux2 -> manylinux2010

**Comment by ehashman** [comment]
There is a bit of a bootstrapping problem in which auditwheel is currently testing against a non-official manylinux2010 image because this one hasn't been released :smile: I'll have an auditwheel release candidate up by the end of the weekend.

**Comment by takluyver** [comment]
Is someone clear what the next steps are for this PR and/or #252?

auditwheel 2.0 is out now, and as @mayeut said, there is a PR (#262) to update to it in this repository. Does that patch need to be applied to this PR?

What are the relative merits of this and #252? @henriquegemignani mentions on that PR that it uses 'devtoolset 7', and 'another approach to the curl issue', but it is (or was?) less ready for merge than this one.

I'm keen to see manylinux2010 in use, and I think the images are the last crucial part of the toolchain. But I lack the low-level knowledge to evaluate the changes in either PR.

**Comment by mayeut** [comment]
Well, I'm not sure anyone's clear one the next steps. I'm also not sure this PR is the place to discuss it either so I'll answer in #179

**Comment by mayeut** [comment]
Superseded by #279 that's been merged-in.

**Pull request closed without merging**